File: code/apps/web/src/lib/chart/mapping.ts
/**
 * Clinical protobuf messages to the models the chart uses.
 *
 * Same boundary discipline as $lib/reception/mapping.ts: every enum is mapped
 * exhaustively, with no `default:` that silently picks a plausible value. In a
 * clinical panel that matters more than it does on a board — a new criticality
 * value read as "low" is a reassurance nobody wrote.
 */
import { timestampDate } from '@bufbuild/protobuf/wkt';
import type { Timestamp } from '@bufbuild/protobuf/wkt';
import {
	AllergyCriticality as WireCriticality,
	AllergyKind as WireAllergyKind,
	AllergyVerification as WireVerification,
	Confidentiality as WireConfidentiality,
	DocumentStatus as WireDocumentStatus,
	Interpretation as WireInterpretation,
	ObservationStatus as WireObservationStatus,
	ProblemStatus as WireProblemStatus,
	SignatureMeaning as WireSignatureMeaning,
	type Allergy as WireAllergy,
	type Document as WireDocument,
	type Observation as WireObservation,
	type Problem as WireProblem
} from '$gen/healthcare/clinical/v1/clinical_pb.js';
import type { ChartDocument, DocumentStatus, SignatureMeaning } from './notes.js';
import {
	presentAllergy,
	presentObservation,
	presentProblem,
	type AllergyKind,
	type Criticality,
	type Interpretation,
	type PresentedAllergy,
	type PresentedObservation,
	type PresentedProblem,
	type ProblemStatus,
	type Verification
} from './safety.js';

function toDate(timestamp: Timestamp | undefined): Date | null {
	return timestamp ? timestampDate(timestamp) : null;
}

const documentStatuses: Record<WireDocumentStatus, DocumentStatus> = {
	[WireDocumentStatus.UNSPECIFIED]: 'unspecified',
	[WireDocumentStatus.DRAFT]: 'draft',
	[WireDocumentStatus.SIGNED]: 'signed',
	[WireDocumentStatus.AMENDED]: 'amended',
	[WireDocumentStatus.ADDENDUM]: 'addendum',
	[WireDocumentStatus.ENTERED_IN_ERROR]: 'entered_in_error'
};

const signatureMeanings: Record<WireSignatureMeaning, SignatureMeaning> = {
	[WireSignatureMeaning.UNSPECIFIED]: 'unspecified',
	[WireSignatureMeaning.AUTHOR]: 'author',
	[WireSignatureMeaning.VERIFIER]: 'verifier',
	[WireSignatureMeaning.COSIGNER]: 'cosigner',
	[WireSignatureMeaning.WITNESS]: 'witness',
	[WireSignatureMeaning.TRANSCRIBER]: 'transcriber'
};

const criticalities: Record<WireCriticality, Criticality> = {
	[WireCriticality.UNSPECIFIED]: 'unspecified',
	[WireCriticality.LOW]: 'low',
	[WireCriticality.HIGH]: 'high',
	// Deliberately not folded into 'low' or 'unspecified'. It is the one value
	// whose whole point is that it is neither.
	[WireCriticality.UNABLE_TO_ASSESS]: 'unable_to_assess'
};

const verifications: Record<WireVerification, Verification> = {
	[WireVerification.UNSPECIFIED]: 'unspecified',
	[WireVerification.UNCONFIRMED]: 'unconfirmed',
	[WireVerification.CONFIRMED]: 'confirmed',
	[WireVerification.REFUTED]: 'refuted',
	[WireVerification.ENTERED_IN_ERROR]: 'entered_in_error'
};

const allergyKinds: Record<WireAllergyKind, AllergyKind> = {
	[WireAllergyKind.UNSPECIFIED]: 'unspecified',
	[WireAllergyKind.ALLERGY]: 'allergy',
	[WireAllergyKind.INTOLERANCE]: 'intolerance'
};

const problemStatuses: Record<WireProblemStatus, ProblemStatus> = {
	[WireProblemStatus.UNSPECIFIED]: 'unspecified',
	[WireProblemStatus.ACTIVE]: 'active',
	[WireProblemStatus.REMISSION]: 'remission',
	[WireProblemStatus.RESOLVED]: 'resolved',
	[WireProblemStatus.INACTIVE]: 'inactive',
	[WireProblemStatus.ENTERED_IN_ERROR]: 'entered_in_error'
};

const interpretations: Record<WireInterpretation, Interpretation> = {
	[WireInterpretation.UNSPECIFIED]: 'unspecified',
	[WireInterpretation.NORMAL]: 'normal',
	[WireInterpretation.HIGH]: 'high',
	[WireInterpretation.LOW]: 'low',
	[WireInterpretation.CRITICAL_HIGH]: 'critical_high',
	[WireInterpretation.CRITICAL_LOW]: 'critical_low',
	[WireInterpretation.ABNORMAL]: 'abnormal',
	[WireInterpretation.UNKNOWN]: 'unknown'
};

const observationStatuses: Record<WireObservationStatus, string> = {
	[WireObservationStatus.UNSPECIFIED]: 'Unknown',
	[WireObservationStatus.REGISTERED]: 'Registered',
	// Reported because a preliminary blood culture at 2am changes treatment,
	// and labelled so nobody mistakes it for the final answer.
	[WireObservationStatus.PRELIMINARY]: 'Preliminary',
	[WireObservationStatus.FINAL]: 'Final',
	[WireObservationStatus.AMENDED]: 'Amended',
	[WireObservationStatus.CANCELLED]: 'Cancelled',
	[WireObservationStatus.ENTERED_IN_ERROR]: 'Entered in error'
};

const confidentialities: Record<WireConfidentiality, string> = {
	[WireConfidentiality.UNSPECIFIED]: 'normal',
	[WireConfidentiality.NORMAL]: 'normal',
	[WireConfidentiality.RESTRICTED]: 'restricted',
	[WireConfidentiality.VERY_RESTRICTED]: 'very_restricted'
};

/** Adapts a wire Document for the chart. */
export function toChartDocument(document: WireDocument): ChartDocument {
	return {
		documentId: document.documentId,
		title: document.title,
		kind: String(document.kind),
		status: documentStatuses[document.status] ?? 'unspecified',
		authoredBy: document.authoredBy,
		createdAt: toDate(document.createdAt) ?? new Date(0),
		updatedAt: toDate(document.updatedAt) ?? new Date(0),
		amendsId: document.amendsId,
		addsToId: document.addsToId,
		changeReason: document.changeReason,
		retractionReason: document.retractionReason,
		dictated: document.dictated,
		signatures: document.signatures.map((signature) => ({
			subjectId: signature.subjectId,
			meaning: signatureMeanings[signature.meaning] ?? 'unspecified',
			signedAt: toDate(signature.signedAt) ?? new Date(0)
		})),
		// Defaults to false when the field is absent rather than true. An
		// unverifiable document treated as intact is the one case where the
		// safe default and the convenient default differ.
		intact: document.intact,
		confidentiality: confidentialities[document.confidentiality] ?? 'normal'
	};
}

/** Adapts a wire Allergy for the safety panel. */
export function toPresentedAllergy(allergy: WireAllergy): PresentedAllergy {
	return presentAllergy({
		allergyId: allergy.allergyId,
		substance: allergy.substance?.display || allergy.substance?.code || 'Unnamed substance',
		kind: allergyKinds[allergy.kind] ?? 'unspecified',
		criticality: criticalities[allergy.criticality] ?? 'unspecified',
		verification: verifications[allergy.verification] ?? 'unspecified',
		reactions: allergy.reactions.map(
			(reaction) => reaction.manifestation?.display || reaction.manifestation?.code || ''
		),
		note: allergy.note
	});
}

/** Adapts a wire Problem for the problem list. */
export function toPresentedProblem(problem: WireProblem): PresentedProblem {
	return presentProblem({
		problemId: problem.problemId,
		code: problem.code?.code ?? '',
		display: problem.code?.display ?? '',
		status: problemStatuses[problem.status] ?? 'unspecified',
		onsetAt: toDate(problem.onsetAt),
		resolvedAt: toDate(problem.resolvedAt),
		note: problem.note
	});
}

/** Adapts a wire Observation for the results panel. */
export function toPresentedObservation(observation: WireObservation): PresentedObservation {
	return presentObservation({
		observationId: observation.observationId,
		display: observation.code?.display || observation.code?.code || 'Result',
		// A quantity of exactly zero is a real measurement, so the presence of
		// the value message decides, not its truthiness.
		value: observation.value ? observation.value.value : null,
		unit: observation.value?.unit ?? '',
		textValue: observation.textValue || observation.codedValue?.display || '',
		interpretation: interpretations[observation.interpretation] ?? 'unspecified',
		interpretationSource: observation.interpretationSource,
		referenceLow: observation.referenceLow,
		referenceHigh: observation.referenceHigh,
		hasReferenceRange: observation.hasReferenceRange,
		referenceText: observation.referenceText,
		effectiveAt: toDate(observation.effectiveAt) ?? new Date(0),
		status: observationStatuses[observation.status] ?? 'Unknown'
	});
}

/** The raw numeric values a trend needs, paired with their ids. */
export function toTrendValues(
	observations: readonly WireObservation[]
): readonly { readonly observationId: string; readonly value: number | null }[] {
	return observations.map((observation) => ({
		observationId: observation.observationId,
		value: observation.value ? observation.value.value : null
	}));
}

File: code/apps/mobile/lib/src/chart/mapping.dart
/// Wire types to the chart's models (UX-W1-02).
///
/// Same discipline as `reception/mapping.dart`, and the reasoning is there in
/// full: `protobuf.dart` decodes an enum value this build has never heard of as
/// the **zero member** and files the real tag under `unknownFields`, so a
/// `default:` arm can never run and the silent failure is the zero member's
/// meaning being applied to something else entirely.
///
/// The stakes here are the highest of the four workspaces. A document status
/// this build cannot read must not fall through to a lifecycle that offers an
/// Edit button, and an allergy criticality it cannot read must not render
/// beside the low-risk rows. Both are handled by an explicit `unrecognised`
/// member whose behaviour is defined to be the cautious one.
library;

import 'package:protobuf/protobuf.dart' as pb;

import '../gen/healthcare/clinical/v1/clinical.pb.dart' as wire;
import 'notes.dart';
import 'safety.dart';

/// Field tags carrying the enums this module reads. Pinned by test.
const int documentStatusField = 9; // clinical.v1.Document.status
const int allergyKindField = 5; // clinical.v1.Allergy.kind
const int allergyCriticalityField = 6; // clinical.v1.Allergy.criticality
const int allergyVerificationField = 7; // clinical.v1.Allergy.verification
const int problemStatusField = 6; // clinical.v1.Problem.status
const int interpretationField = 12; // clinical.v1.Observation.interpretation

/// True when [message] carried a value for [tag] this build could not read.
bool sentUnknownValueFor(pb.GeneratedMessage message, int tag) =>
    message.unknownFields.hasField(tag);

// ------------------------------------------------------------------- notes

DocumentStatus documentStatusOf(wire.Document document) {
  if (sentUnknownValueFor(document, documentStatusField)) {
    return DocumentStatus.unrecognised;
  }
  return switch (document.status) {
    wire.DocumentStatus.DOCUMENT_STATUS_DRAFT => DocumentStatus.draft,
    wire.DocumentStatus.DOCUMENT_STATUS_SIGNED => DocumentStatus.signed,
    wire.DocumentStatus.DOCUMENT_STATUS_AMENDED => DocumentStatus.amended,
    wire.DocumentStatus.DOCUMENT_STATUS_ADDENDUM => DocumentStatus.addendum,
    wire.DocumentStatus.DOCUMENT_STATUS_ENTERED_IN_ERROR =>
      DocumentStatus.enteredInError,
    wire.DocumentStatus.DOCUMENT_STATUS_UNSPECIFIED => DocumentStatus.unspecified,
    _ => DocumentStatus.unrecognised,
  };
}

/// Signature meaning, wire to model.
///
/// Takes the signature rather than the document because the tag lives on the
/// signature, and because a document with one unreadable signature among four
/// should not lose the other three.
SignatureMeaning signatureMeaningOf(wire.Signature signature, {int tag = 2}) {
  if (sentUnknownValueFor(signature, tag)) {
    // Not "author". An unreadable meaning must never finalise a note, because
    // finalising is what makes it un-editable and clinically attributed.
    return SignatureMeaning.unrecognised;
  }
  return switch (signature.meaning) {
    wire.SignatureMeaning.SIGNATURE_MEANING_AUTHOR => SignatureMeaning.author,
    wire.SignatureMeaning.SIGNATURE_MEANING_VERIFIER => SignatureMeaning.verifier,
    wire.SignatureMeaning.SIGNATURE_MEANING_COSIGNER => SignatureMeaning.cosigner,
    wire.SignatureMeaning.SIGNATURE_MEANING_WITNESS => SignatureMeaning.witness,
    wire.SignatureMeaning.SIGNATURE_MEANING_TRANSCRIBER =>
      SignatureMeaning.transcriber,
    wire.SignatureMeaning.SIGNATURE_MEANING_UNSPECIFIED =>
      SignatureMeaning.unspecified,
    _ => SignatureMeaning.unrecognised,
  };
}

ChartDocument documentOf(wire.Document document) => ChartDocument(
      documentId: document.documentId,
      title: document.title,
      kind: document.kind.name,
      status: documentStatusOf(document),
      authoredBy: document.authoredBy,
      createdAt: document.createdAt.toDateTime().toUtc(),
      amendsId: document.amendsId,
      addsToId: document.addsToId,
      changeReason: document.changeReason,
      retractionReason: document.retractionReason,
      dictated: document.dictated,
      signatures: [
        for (final s in document.signatures)
          DocumentSignature(
            subjectId: s.subjectId,
            meaning: signatureMeaningOf(s),
            signedAt: s.signedAt.toDateTime().toUtc(),
          ),
      ],
      intact: document.intact,
      confidentiality: document.confidentiality.name,
    );

// ---------------------------------------------------------------- allergies

Criticality criticalityOf(wire.Allergy allergy) {
  if (sentUnknownValueFor(allergy, allergyCriticalityField)) {
    return Criticality.unrecognised;
  }
  return switch (allergy.criticality) {
    wire.AllergyCriticality.ALLERGY_CRITICALITY_LOW => Criticality.low,
    wire.AllergyCriticality.ALLERGY_CRITICALITY_HIGH => Criticality.high,
    wire.AllergyCriticality.ALLERGY_CRITICALITY_UNABLE_TO_ASSESS =>
      Criticality.unableToAssess,
    wire.AllergyCriticality.ALLERGY_CRITICALITY_UNSPECIFIED =>
      Criticality.unspecified,
    _ => Criticality.unrecognised,
  };
}

Verification verificationOf(wire.Allergy allergy) {
  if (sentUnknownValueFor(allergy, allergyVerificationField)) {
    // Not "refuted": that would move a live allergy into the historical
    // section and out of a prescriber's way, which is the worst available
    // answer.
    return Verification.unrecognised;
  }
  return switch (allergy.verification) {
    wire.AllergyVerification.ALLERGY_VERIFICATION_UNCONFIRMED =>
      Verification.unconfirmed,
    wire.AllergyVerification.ALLERGY_VERIFICATION_CONFIRMED =>
      Verification.confirmed,
    wire.AllergyVerification.ALLERGY_VERIFICATION_REFUTED => Verification.refuted,
    wire.AllergyVerification.ALLERGY_VERIFICATION_ENTERED_IN_ERROR =>
      Verification.enteredInError,
    wire.AllergyVerification.ALLERGY_VERIFICATION_UNSPECIFIED =>
      Verification.unspecified,
    _ => Verification.unrecognised,
  };
}

AllergyKind allergyKindOf(wire.Allergy allergy) {
  if (sentUnknownValueFor(allergy, allergyKindField)) {
    return AllergyKind.unrecognised;
  }
  return switch (allergy.kind) {
    wire.AllergyKind.ALLERGY_KIND_ALLERGY => AllergyKind.allergy,
    wire.AllergyKind.ALLERGY_KIND_INTOLERANCE => AllergyKind.intolerance,
    wire.AllergyKind.ALLERGY_KIND_UNSPECIFIED => AllergyKind.unspecified,
    _ => AllergyKind.unrecognised,
  };
}

PresentedAllergy allergyOf(wire.Allergy allergy) => presentAllergy(
      allergyId: allergy.allergyId,
      substance: allergy.substance.display.isNotEmpty
          ? allergy.substance.display
          // A substance with no display text still has to name something: the
          // code is worse to read and far better than a blank row on the panel
          // a prescriber checks.
          : allergy.substance.code,
      kind: allergyKindOf(allergy),
      criticality: criticalityOf(allergy),
      verification: verificationOf(allergy),
      reactions: [
        for (final r in allergy.reactions)
          r.manifestation.display.isNotEmpty
              ? r.manifestation.display
              : r.manifestation.code,
      ],
      note: allergy.note,
    );

// ----------------------------------------------------------------- problems

ProblemStatus problemStatusOf(wire.Problem problem) {
  if (sentUnknownValueFor(problem, problemStatusField)) {
    return ProblemStatus.unrecognised;
  }
  return switch (problem.status) {
    wire.ProblemStatus.PROBLEM_STATUS_ACTIVE => ProblemStatus.active,
    wire.ProblemStatus.PROBLEM_STATUS_REMISSION => ProblemStatus.remission,
    wire.ProblemStatus.PROBLEM_STATUS_RESOLVED => ProblemStatus.resolved,
    wire.ProblemStatus.PROBLEM_STATUS_INACTIVE => ProblemStatus.inactive,
    wire.ProblemStatus.PROBLEM_STATUS_ENTERED_IN_ERROR =>
      ProblemStatus.enteredInError,
    wire.ProblemStatus.PROBLEM_STATUS_UNSPECIFIED => ProblemStatus.unspecified,
    _ => ProblemStatus.unrecognised,
  };
}

PresentedProblem problemOf(wire.Problem problem) => presentProblem(
      problemId: problem.problemId,
      code: problem.code.code,
      display: problem.code.display.isNotEmpty
          ? problem.code.display
          : problem.code.code,
      status: problemStatusOf(problem),
      onsetAt:
          problem.hasOnsetAt() ? problem.onsetAt.toDateTime().toUtc() : null,
      resolvedAt: problem.hasResolvedAt()
          ? problem.resolvedAt.toDateTime().toUtc()
          : null,
    );

// ------------------------------------------------------------- observations

Interpretation interpretationOf(wire.Observation observation) {
  if (sentUnknownValueFor(observation, interpretationField)) {
    // Not "normal". The zero member here means "nobody interpreted this",
    // which is already the careful reading; an unreadable one says so in its
    // own words so the label does not claim a laboratory verdict.
    return Interpretation.unrecognised;
  }
  return switch (observation.interpretation) {
    wire.Interpretation.INTERPRETATION_NORMAL => Interpretation.normal,
    wire.Interpretation.INTERPRETATION_HIGH => Interpretation.high,
    wire.Interpretation.INTERPRETATION_LOW => Interpretation.low,
    wire.Interpretation.INTERPRETATION_CRITICAL_HIGH => Interpretation.criticalHigh,
    wire.Interpretation.INTERPRETATION_CRITICAL_LOW => Interpretation.criticalLow,
    wire.Interpretation.INTERPRETATION_ABNORMAL => Interpretation.abnormal,
    wire.Interpretation.INTERPRETATION_UNKNOWN => Interpretation.unknown,
    wire.Interpretation.INTERPRETATION_UNSPECIFIED => Interpretation.unspecified,
    _ => Interpretation.unrecognised,
  };
}

PresentedObservation observationOf(wire.Observation observation) =>
    presentObservation(
      observationId: observation.observationId,
      display: observation.code.display.isNotEmpty
          ? observation.code.display
          : observation.code.code,
      effectiveAt: observation.effectiveAt.toDateTime().toUtc(),
      interpretation: interpretationOf(observation),
      // hasValue rather than a non-zero check: a measured zero is a result, and
      // a platelet count of zero read as "no numeric value" is a result that
      // disappears off the trend it most needs to be on.
      value: observation.hasValue() ? observation.value.value : null,
      unit: observation.hasValue() ? observation.value.unit : '',
      textValue: observation.textValue.isNotEmpty
          ? observation.textValue
          : observation.codedValue.display,
      interpretationSource: observation.interpretationSource,
      referenceLow:
          observation.hasReferenceRange ? observation.referenceLow : null,
      referenceHigh:
          observation.hasReferenceRange ? observation.referenceHigh : null,
      referenceText: observation.referenceText,
      status: observation.status.name,
    );
